internal/storage/sql/common/rule.go

```diff
@@ -286,65 +364,143 @@ func (s *Store) CountRules(ctx context.Context, namespaceKey, flagKey string) (u
+			Id:              uuid.Must(uuid.NewV4()).String(),
+			NamespaceKey:    r.NamespaceKey,
+			FlagKey:         r.FlagKey,
+			Rank:            r.Rank,
+			SegmentOperator: r.SegmentOperator,
+			CreatedAt:       now,
+			UpdatedAt:       now,
 		}
 	)
 
+	// Force segment operator to be OR when `segmentKeys` length is 1.
+	if len(segmentKeys) == 1 {
+		rule.SegmentOperator = flipt.SegmentOperator_OR_SEGMENT_OPERATOR
+	}
+
+	tx, err := s.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			_ = tx.Rollback()
+		}
+	}()
+
 	if _, err := s.builder.
 		Insert("rules").
-		Columns("id", "namespace_key", "flag_key", "segment_key", "\"rank\"", "created_at", "updated_at").
+		RunWith(tx).
+		Columns("id", "namespace_key", "flag_key", "\"rank\"", "segment_operator", "created_at", "updated_at").
 		Values(
 			rule.Id,
 			rule.NamespaceKey,
 			rule.FlagKey,
-			rule.SegmentKey,
 			rule.Rank,
+			rule.SegmentOperator,
 			&fliptsql.Timestamp{Timestamp: rule.CreatedAt},
-			&fliptsql.Timestamp{Timestamp: rule.UpdatedAt}).
+			&fliptsql.Timestamp{Timestamp: rule.UpdatedAt},
+		).
 		ExecContext(ctx); err != nil {
 		return nil, err
 	}
 
-	return rule, nil
+	for _, segmentKey := range segmentKeys {
+		if _, err := s.builder.
+			Insert("rule_segments").
+			RunWith(tx).
+			Columns("rule_id", "namespace_key", "segment_key").
+			Values(
+				rule.Id,
+				rule.NamespaceKey,
+				segmentKey,
+			).
+			ExecContext(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(segmentKeys) == 1 {
+		rule.SegmentKey = segmentKeys[0]
+	} else {
+		rule.SegmentKeys = segmentKeys
+	}
+
+	return rule, tx.Commit()
 }
 
 // UpdateRule updates an existing rule
-func (s *Store) UpdateRule(ctx context.Context, r *flipt.UpdateRuleRequest) (*flipt.Rule, error) {
+func (s *Store) UpdateRule(ctx context.Context, r *flipt.UpdateRuleRequest) (_ *flipt.Rule, err error) {
+	segmentKeys := sanitizeSegmentKeys(r.GetSegmentKey(), r.GetSegmentKeys())
+
 	if r.NamespaceKey == "" {
 		r.NamespaceKey = storage.DefaultNamespace
 	}
 
-	query := s.builder.Update("rules").
-		Set("segment_key", r.SegmentKey).
-		Set("updated_at", &fliptsql.Timestamp{Timestamp: timestamppb.Now()}).
-		Where(sq.And{sq.Eq{"id": r.Id}, sq.Eq{"namespace_key": r.NamespaceKey}, sq.Eq{"flag_key": r.FlagKey}})
-
-	res, err := query.ExecContext(ctx)
+	tx, err := s.db.Begin()
 	if err != nil {
 		return nil, err
 	}
 
-	count, err := res.RowsAffected()
+	defer func() {
+		if err != nil {
+			_ = tx.Rollback()
+		}
+	}()
+
+	var segmentOperator = r.SegmentOperator
+	if len(segmentKeys) == 1 {
+		segmentOperator = flipt.SegmentOperator_OR_SEGMENT_OPERATOR
+	}
+
+	// Set segment operator.
+	_, err = s.builder.Update("rules").
+		RunWith(tx).
+		Set("segment_operator", segmentOperator).
+		Set("updated_at", &fliptsql.Timestamp{Timestamp: timestamppb.Now()}).
+		Where(sq.And{sq.Eq{"id": r.Id}, sq.Eq{"namespace_key": r.NamespaceKey}, sq.Eq{"flag_key": r.FlagKey}}).
```

Comment:
Another small optimization. `sq.And` and `sq.Eq` are maps. So each time they're stated we're allocating a map. Squirell with implicity AND all the k/v equalities in a single `sq.Eq` map.
```suggestion
		Where(sq.Eq{"id": r.Id, "namespace_key": r.NamespaceKey, "flag_key": r.FlagKey}).
```